### PR TITLE
Add XMC4xxx gpio drivers

### PR DESCRIPTION
**Pull request by talih0**
Interrupt support will be added after interrupt controller (called ERU on xmc4xxx) driver.

### CODE REVIEW

File: dts/bindings/gpio/infineon,xmc4xxx-gpio.yaml [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+compatible: "infineon,xmc4xxx-gpio"
+
+include: [gpio-controller.yaml, base.yaml]
+
+properties:
+    reg:
+      required: true
+
+    label:
```
Comment (galak) [inline dts/bindings/gpio/infineon,xmc4xxx-gpio.yaml:11]: remove `label` and from .dts(i) files.

File: boards/arm/xmc45_relax_kit/xmc45_relax_kit.dts [hunk 2]
```diff
@@ -15,6 +15,20 @@
 	compatible = "infineon,xm4500-relax-kit", "infineon,xmc4500",
 						"infineon,xmc4xxx";
 
+	aliases {
+		led0 = &led1;
```
Comment (mbolivar-nordic) [inline boards/arm/xmc45_relax_kit/xmc45_relax_kit.dts:19]: ???

Comment (talih0) [reply boards/arm/xmc45_relax_kit/xmc45_relax_kit.dts:19]: It's an alias for running the blinky sample:
https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/basic/blinky/src/main.c#L14

The leds are labeled led1 and led2 in the relax kit documentation:
https://www.infineon.com/dgdl/Board_Users_Manual_XMC4500_Relax_Kit-V1_R1.2_released.pdf?fileId=db3a30433acf32c9013adf6b97b112f9

Comment (mbolivar-nordic) [reply boards/arm/xmc45_relax_kit/xmc45_relax_kit.dts:19]: This at least deserves some source code comments IMO.

Comment (talih0) [reply boards/arm/xmc45_relax_kit/xmc45_relax_kit.dts:19]: will update.

Comment (mbolivar-nordic) [reply boards/arm/xmc45_relax_kit/xmc45_relax_kit.dts:19]: Thanks! I agree this is resolved.

**Comment by talih0** [comment]
I added an extra field in the bindings. It will be useful for the interrupt controller:

    @@ dts/bindings/gpio/infineon,xmc4xxx-gpio.yaml (new)
     +    "#gpio-cells":
     +      const: 2
     +
    -+    port-id:
    -+      description: Maps port name to an index. gpioK maps to an index K.
    -+      type: int
    -+      required: true

**Review by gmarull — COMMENTED** [review]
can you confirm the driver passes `tests/drivers/gpio`?

File: drivers/gpio/Kconfig.xmc4xxx [hunk 3]
```diff
@@ -0,0 +1,12 @@
+# XMC4XXX GPIO configuration
+#
+# Copyright (c) 2022 Schlumberger
+# SPDX-License-Identifier: Apache-2.0
+
+config GPIO_XMC4XXX
+	bool "XMC4XXX GPIO driver"
+	default y
+	depends on DT_HAS_INFINEON_XMC4XXX_GPIO_ENABLED
+	depends on SOC_SERIES_XMC_4XXX
```
[the remark this answers is not in the document]
Comment (talih0) [reply drivers/gpio/Kconfig.xmc4xxx:10]: > this is now redundant

just to clarify, did you mean the SOC_SERIES_XMC_4XXX symbol?
@galak suggested to use the DT_HAS_xx_ENABLED symbol

File: boards/arm/xmc45_relax_kit/xmc45_relax_kit.dts [hunk 4]
```diff
@@ -33,3 +33,27 @@
...
+	status = "okay";
+};
+
+&gpio4 {
+	status = "okay";
+};
+
+&gpio5 {
+	status = "okay";
+};
```
[the remark this answers is not in the document]
Comment (talih0) [reply boards/arm/xmc45_relax_kit/xmc45_relax_kit.dts:74]: good point, will remove the unused ones.

File: drivers/gpio/Kconfig.xmc4xxx (hunk 3 above)
[the remark this answers is not in the document]
Comment (gmarull) [reply drivers/gpio/Kconfig.xmc4xxx:10]: yes, the `SOC_` symbol, the DT_ being `y` means you're on a valid SoC when selecting it

**Review by gmarull — DISMISSED** [review]

**Comment by talih0** [comment]
ping

**Comment by talih0** [comment]
I rebased to fix merge conflicts.

**Review by gmarull — APPROVED** [review]

**Review by cfriedt — APPROVED** [review]

**Pull request merged**